docs/web_configuration.md
# Web Configuration

The application can be configured using command-line parameters, or environment variables

The following options are available:

* `HTTP_PORT` / `--http-port` - the HTTP address (default: 8000)
* `HTTP_ADDRESS` / `--http-address` - the HTTP address (default: 127.0.0.1)
* `DATA_DIR` - the data directory root (default: ".")
* `SURREAL_DB_CONNECTION` - the surreal DB connection (default: "ws://localhost:8800") - set to `rocksdb://data/surreal` for embedded mode
* `BITCOIN_NETWORK` - bitcoin network to use (default: testnet), possible values: `mainnet`, `regtest`, `testnet4` and `testnet`
* `RUST_LOG` - the log level, e.g.: info, trace, debug, error (default: error)
* `NOSTR_RELAY` - nostr relay endpoint (default: ws://localhost:8080)
* `MINT_URL` - cashu mint endpoint (default: https://moksha.minibill.tech)
* `JOB_RUNNER_INITIAL_DELAY_SECONDS` - initial delay until cron jobs run (default: 1)
* `JOB_RUNNER_CHECK_INTERVAL_SECONDS` - interval in which cron jobs run (default: 600)
* `FRONTEND_URL_PATH` - default path to serve the frontend from (default: /)
* `FRONTEND_SERVE_FOLDER` - folder where the static frontend is served from (default: ./frontend)
* `LAUNCH_FRONTEND_AT_STARTUP` - open the frontend in a browser on startup (default: false)
* `ESPLORA_BASE_URL` - The base url of the esplora URL for payment checks (default: https://esplora.minibill.tech)

## Example

```bash
RUST_LOG=info cargo run -- --http-port 8001

RUST_LOG=info HTTP_PORT=8001 cargo run
```



docs/wasm_configuration.md
# WASM Configuration

The application can be configured using the `Config` struct.

```rust
pub struct Config {
    pub bitcoin_network: String,
    pub nostr_relays: Vec<String>,
    pub esplora_base_url: String,
    pub surreal_db_connection: String,
    pub data_dir: String,
    pub job_runner_initial_delay_seconds: u32,
    pub job_runner_check_interval_seconds: u32,
}
```

It contains the following options:

* `bitcoin_network` - bitcoin network to use, possible values: `mainnet`, `regtest`, `testnet4` and `testnet`
* `esplora_base_url` - bitcoin explorer base url for payment checks
* `nostr_relays` - the nostr relay endpoints
* `surreal_db_connection` - the surreal DB connection
* `data_dir` - the data directory root - not used on the Web
* `job_runner_initial_delay_seconds` - initial delay until cron jobs run
* `job_runner_check_interval_seconds` - interval in which cron jobs run

## Example

```javascript
    let config = {
        bitcoin_network: "testnet",
        esplora_base_url: "https://esplora.minibill.tech",
        nostr_relays: ["wss://bitcr-cloud-run-05-550030097098.europe-west1.run.app"],
        surreal_db_connection: "indxdb://default",
        data_dir: ".",
        job_runner_initial_delay_seconds: 1,
        job_runner_check_interval_seconds: 600,
    };
```



CHANGELOG.md
# 0.3.12

* impl Default for BillParticipant and BillAnonParticipant
* Expose Receiver type for the PushApi trait
* Fix SurrealDB Memory leak for WASM

# 0.3.11

* Add LICENSE to crates
* Remove the `bcr-ebill-web` crate
* Use a list of Nostr relays everywhere, instead of a single one, including in the config
* Add Blank Endorse Bill data model implementation
    * Rename `IdentityPublicData` to `BillIdentParticipant`
        * same for `LightIdentityPublicData`
    * Introduce the concept of `BillParticipant`, with the variants `Ident` and `Anon`
        * `Anon` includes a `BillAnonParticipant`
        * `Ident` includes a `BillIdentParticipant`
    * Use `BillParticipant` in parts of the bill where a participant can be anonymous
* Add the possibility to add anonymous contacts
    * They only have Node Id, Name and E-Mail as fields
    * E-Mail is optional
    * This changes the data model for contacts, `email` and `postal_address` are now optional
        * Additional validation rules ensure the fields can only be set for non-anon contacts
    * Adds an endpoint `Api.contact().deanonymize()` to de-anonymize a contact by adding all necessary fields for a personal, or company contact
        * It takes the same payload as creating a contact and fails for non-anon contacts
* Add the possibility to add an anonymous identity
    * They only have Node Id, (nick)name and E-Mail as fields
    * E-Mail is optional
    * Adds an endpoint `Api.identity().deanonymize()` to de-anonymize an identity by adding all necessary fields for a personal identity
        * It takes the same payload as creating an identity and fails for a non-anon identity
    * Anon identity can't issue bills, or create a company
* Add the possibility to issue and endorse blank
    * New `Api.bill()` methods for blank endorsements
      * `issue_blank`
      * `offer_to_sell_blank`
      * `endorse_bill_blank`
      * `mint_bill_blank`
    * Can issue (non-self-drafted) blank bills (payee is anon)
    * Can endorse/mint/offer to sell to anon endorsee/mint/buyer
    * If caller of a bill action is anonymous in the bill, any action they take stay anonymous (e.g. endorse)
* Add endpoint to check payment for singular bill
    * `Api.bill().check_payment_for_bill(id)`
* Fix TS type for identity detail
* Return identity on `create` and `deanonymize` identity for consistency

# 0.3.10

* Change default testnet block explorer to `https://esplora.minibill.tech`
* Add LICENSE to npm package
* Reduce size of the WASM binary
* Fix a small issue, where bills were recalculated instead of taken from cache, once their payment/sell/recourse/accept requests expired
* Change behaviour of request to pay
    * it's now possible to req to pay before the maturity date
    * The actual payment expiry still only happens 2 workdays after the end of the maturity date,
    or end of the req to pay end of day if that was after the maturity date
    * the `request_to_pay_timed_out` flag is set after payment expired, not after the req to pay expired
    * The waiting state for payment is only active during the req to pay (while it's blocked)
      * Afterwards, the bill is not blocked anymore, can still be rejected to pay and paid
      * But recourse is only possible after the payment expired (after maturity date)
    * An expired req to pay, which expired before the maturity date does not show up in `past_payments`

# 0.3.9

* Add possibility to use a local regtest esplora setup for payment

# 0.3.8

* Add `esplora_base_url` as config parameter to be able to use a custom esplora based block explorer
* Add `node_ids` filter to `list` notifications endpoint
* Fixed an issue where events weren't propagated if no one was subscribed to the push notifications
* Run payment checks on startup

# 0.3.7

* Fix request recourse to accept validation - does not require a request to accept anymore

# 0.3.6

* Add validation for maturity date
* Add docs for testing
* Fix reject to accept not showing correctly without req to accept
* Add endpoint `clear_bill_cache` to clear the bill cache

# 0.3.5

* Properly propagate and log errors when getting a file (e.g. an avatar)
* Several fixes to recourse bill action validation
* Add in-depth tests for bill validation
* Fix not checking contact for company files

# 0.3.4

* Add in-depth tests for bill validation
* Add recourse reason to `Recourse` block data
    * (breaks existing persisted bills, if they had a recourse block)
* Added `has_requested_funds` flag to `BillStatusWeb`, indicating the caller has requested funds (req to pay, req to recourse, offer to sell) at some point
* Added `past_payments` endpoint to `Api.bill()`, which returns data about past payments and payment requests where the caller was the beneficiary

# 0.3.3

* Use Nip-04 as a default for Nostr communication
* Add incoming bill validation
* Add block data validation
* Add bill action validation for incoming blocks
* Add signer verification for incoming blocks
* Add recourse reason to `RequestRecourse` block data
    * (breaks existing persisted bills, if they had a request recourse block)
* Move bill validation logic to `bcr-ebill-core`

# 0.3.2

* Fixed `request_to_accept` calling the correct action
* Multi-identity Nostr consumer and currently-active-identity-sending
* Added more thorough logging, especially debug logging
* Expose Error types to TS
* Use string for `log_level` in config

# 0.3.1

* Persist active Identity to DB for WASM
* Change indexed-db name to "data"
* Use a different indexed-db collection for files, named "files"
* Create a new indexeddb database connection for each query to avoid transaction overlapping
* Removed timezone db api
* Persist base64 string instead of bytes for images, for more efficiency
* Added Retry-sending for Nostr block events
* Added block propagation via Nostr
* Added a caching layer for bills, heavily improving performance
* Added `error` logs for all errors returned from the API for the WASM version

crates/bcr-ebill-web/src/config.rs
use clap::Parser;

/// Configuration for the bitcredit application
/// Allows to set the ports and addresses for the network connections
/// either via command line or environment variables
#[derive(Parser, Clone)]
#[command(version, about, long_about = None)]
pub struct Config {
    #[arg(default_value_t = 8000, long, env = "HTTP_PORT")]
    pub http_port: u16,
    #[arg(default_value_t = String::from("127.0.0.1"), long, env = "HTTP_ADDRESS")]
    pub http_address: String,
    #[arg(default_value_t = String::from("."), long, env = "DATA_DIR")]
    pub data_dir: String,
    #[arg(default_value_t = String::from("ws://localhost:8800"), long, env = "SURREAL_DB_CONNECTION")]
    pub surreal_db_connection: String,
    #[arg(default_value_t = String::from("testnet"),  long, env = "BITCOIN_NETWORK")]
    pub bitcoin_network: String,
    #[arg(default_value_t = String::from("ws://localhost:8080"), long, env = "NOSTR_RELAY")]
    pub nostr_relay: String,
    #[arg(default_value_t = String::from("https://moksha.minibill.tech"), long, env = "MINT_URL")]
    pub mint_url: String,
    #[arg(default_value_t = 1, long, env = "JOB_RUNNER_INITIAL_DELAY_SECONDS")]
    pub job_runner_initial_delay_seconds: u64,
    #[arg(default_value_t = 600, long, env = "JOB_RUNNER_CHECK_INTERVAL_SECONDS")]
    pub job_runner_check_interval_seconds: u64,
    #[arg(default_value_t = String::from("frontend"), long, env = "FRONTEND_SERVE_FOLDER")]
    pub frontend_serve_folder: String,
    #[arg(default_value_t = String::from("/"), long, env = "FRONTEND_URL_PATH")]
    pub frontend_url_path: String,
    #[arg(default_value_t = false, long, env = "LAUNCH_FRONTEND_AT_STARTUP")]
    pub launch_frontend_at_startup: bool,
    #[arg(default_value_t = String::from("https://esplora.minibill.tech"), long, env = "ESPLORA_BASE_URL")]
    pub esplora_base_url: String,
}

impl Config {
    pub fn http_listen_url(&self) -> String {
        format!("http://{}:{}", self.http_address, self.http_port)
    }
}
